beetsplug/subsonicplaylist.py

```diff
@@ -0,0 +1,138 @@
+                'password': ''
+            }
+        )
+        self.config['password'].redact = True
+
+    def update_tags(self, playlist_dict, lib):
+        with lib.transaction():
+            for query, playlist_tag in playlist_dict.items():
+                query = AndQuery([SubstringQuery("artist", query[0]),
+                                  SubstringQuery("album", query[1]),
+                                  SubstringQuery("title", query[2])])
+                items = lib.items(query)
+                if not items:
+                    self._log.warn(u"{} | track not found ({})", playlist_tag,
+                                   query)
+                    continue
+                for item in items:
+                    item.subsonic_playlist = playlist_tag
+                    item.try_sync(write=True, move=False)
+
+    def get_playlist(self, playlist_id):
+        xml = self.send('getPlaylist', {'id': playlist_id}).text
+        playlist = ET.fromstring(xml)[0]
+        if playlist.attrib.get('code', '200') != '200':
+            alt_error = 'error getting playlist, but no error message found'
+            self._log.warn(playlist.attrib.get('message', alt_error))
+            return
+
+        name = playlist.attrib.get('name', 'undefined')
+        tracks = [(t.attrib['artist'], t.attrib['album'], t.attrib['title'])
+                  for t in playlist]
+        return name, tracks
+
+    def commands(self):
+        def build_playlist(lib, opts, args):
+            self.config.set_args(opts)
+            ids = self.config['playlist_ids'].as_str_seq()
+            if len(self.config['playlist_names'].as_str_seq()) > 0:
+                playlists = ET.fromstring(self.send('getPlaylists').text)[
+                    0]
+                if playlists.attrib.get('code', '200') != '200':
+                    alt_error = 'error getting playlists,' \
+                                ' but no erro message found'
+                    self._log.warn(
+                        playlists.attrib.get('message', alt_error))
+                    return
+                for name in self.config['playlist_names'].as_str_seq():
+                    for playlist in playlists:
+                        if name == playlist.attrib['name']:
+                            ids.append(playlist.attrib['id'])
+            playlist_dict = dict()
+            for playlist_id in ids:
+                name, tracks = self.get_playlist(playlist_id)
+                for track in tracks:
+                    if track not in playlist_dict:
+                        playlist_dict[track] = ';'
+                    playlist_dict[track] += name + ';'
+            # delete old tags
+            if self.config['delete']:
+                for item in lib.items('subsonic_playlist:";"'):
+                    item.update({'subsonic_playlist': ''})
+                    with lib.transaction():
+                        item.try_sync(write=True, move=False)
```

Comment:
Since the new metadata is being saved anyway, is this necessary?

Reply:
this is for songs removed from a playlist as the aren't contained in the playlist_dict anymore. But I should be able to optimize it, so that it doesn't updates songs contained in the playlist_dict